### PR TITLE
Feature/issue85/add refreshable option

### PR DESCRIPTION
**Pull request by annapiktas**
Adding a `Refreshable` protocol similar to the existing `Completable` & `Dismissible` protocols. Each thing that conforms to `Refreshable` is responsible for how to refresh itself e.g. a refreshable screen needs pulling upwards to trigger the refresh.
Refresh tests were added to Table & Collection View Tests. This is #85 TABTestKit enhancement.

### CODE REVIEW

File: Example/TABTestKit.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -649,6 +649,7 @@
 				CLANG_ANALYZER_NUMBER_OBJECT_CONVERSION = YES_AGGRESSIVE;
 				CLANG_ENABLE_MODULES = YES;
 				CLANG_WARN_DOCUMENTATION_COMMENTS = YES;
+				"CODE_SIGN_IDENTITY[sdk=macosx*]" = "Apple Development";
```
Comment (zacoid55) [inline Example/TABTestKit.xcodeproj/project.pbxproj:652]: This might the reason the builds are failing on Travis. Is there a reason you added this in?

File: Example/TABTestKit.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -307,7 +307,6 @@
 					};
 					70FEA77B1F6AC29400633C5C = {
 						CreatedOnToolsVersion = 8.3.3;
-						DevelopmentTeam = PR289YX2Y5;
```
Comment (zacoid55) [inline Example/TABTestKit.xcodeproj/project.pbxproj:310]: @annapiktas Apologies for the mess around but do you mind if you revert this change and the two below, please? As this wasn't the culprit for the build issue, I'm not sure what taking them away will do in the long run

Comment (annapiktas) [reply Example/TABTestKit.xcodeproj/project.pbxproj]: Np, thanks for catching this one.

File: Example/TABTestKit_ExampleUITests/CollectionViewScreen.swift [hunk 3]
```diff
@@ -17,3 +17,11 @@ struct CollectionViewScreen: ScrollableScreen {
   lazy var lastCell = trait.cell(matchingID: "CollectionCell 49")
   
 }
+
+extension CollectionViewScreen: Refreshable {
+    
+  func refresh() {
+    // Pull the collection view down to refresh
+    trait.scroll(.upwards)
```
Comment (zacoid55) [inline Example/TABTestKit_ExampleUITests/CollectionViewScreen.swift:25]: Do we need to make sure that you're at the top of the screen? There could be a chance that we're in a scrolled position and this won't work. Same for the other one below

File: TABTestKit.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -7,6 +7,7 @@
 	objects = {
 
 /* Begin PBXBuildFile section */
+		2DEC7C4924D0822D00545C45 /* Refreshable.swift in Sources */ = {isa = PBXBuildFile; fileRef = 2DEC7C4824D0822D00545C45 /* Refreshable.swift */; };
```
[the remark this answers is not in the document]
Comment (annapiktas) [reply TABTestKit.xcodeproj/project.pbxproj:10]: Thx, I leant it hard way, the travis was initially failing and not giving out much in the error message until Zac explain and show me how to add the new ref for Carthage project.

File: TABTestKit/Classes/Contexts/InteractionContext.swift [hunk 5]
```diff
@@ -72,4 +72,8 @@ public extension InteractionContext {
 		element.adjust(to: newValue)
 	}
 	
+	func refresh(_ refreshableThing: Refreshable) {
+	  refreshableThing.refresh()
```
[the remark this answers is not in the document]
Comment (annapiktas) [reply TABTestKit/Classes/Contexts/InteractionContext.swift:76]: You are right, I tried to stick with whatever indentation was in the file already but will try to get this right.

File: TABTestKit/Classes/Contexts/InteractionContext.swift [hunk 6]
```diff
@@ -53,27 +53,31 @@ public extension InteractionContext {
...
-				element.scroll(direction)
-			} while numberOfTries <= maxTries
-			XCTFail("Ran of out tries (\(maxTries)) waiting for element to become \(value)")
+		var numberOfTries = 0
+		repeat {
+			guard !(otherElement.value == value) else { return }
+			numberOfTries += 1
+			element.scroll(direction)
+		} while numberOfTries <= maxTries
+		XCTFail("Ran of out tries (\(maxTries)) waiting for element to become \(value)")
```
Comment (roger-tan) [inline TABTestKit/Classes/Contexts/InteractionContext.swift:64]:
```suggestion
        var numberOfTries = 0
        repeat {
            guard !(otherElement.value == value) else { return }
            numberOfTries += 1
            element.scroll(direction)
        } while numberOfTries <= maxTries
        XCTFail("Ran of out tries (\(maxTries)) waiting for element to become \(value)")
```

File: TABTestKit/Classes/Contexts/InteractionContext.swift [hunk 7]
```diff
@@ -53,27 +53,31 @@ public extension InteractionContext {
...
 	}
 	
 	func value<ElementWithValue: Element & ValueRepresentable>(of element: ElementWithValue, is expectedValue: ElementWithValue.Value) {
 		XCTAssertTrue(element.underlyingXCUIElement.wait(for: element.value == expectedValue), "Element did not have the right value before timing out! Expected: \(expectedValue), actual: \(element.value)")
 	}
 	
 	func label(of element: Element, is expectedLabel: String) {
-	        XCTAssertTrue(element.underlyingXCUIElement.wait(for: element.label == expectedLabel), "Element did not have the right label before timing out! Expected: \(expectedLabel), actual: \(element.label)")
+		XCTAssertTrue(element.underlyingXCUIElement.wait(for: element.label == expectedLabel), "Element did not have the right label before timing out! Expected: \(expectedLabel), actual: \(element.label)")
```
Comment (roger-tan) [inline TABTestKit/Classes/Contexts/InteractionContext.swift:72]:
```suggestion
        XCTAssertTrue(element.underlyingXCUIElement.wait(for: element.label == expectedLabel), "Element did not have the right label before timing out! Expected: \(expectedLabel), actual: \(element.label)")
```

File: TABTestKit/Classes/Contexts/InteractionContext.swift [hunk 8]
```diff
@@ -53,27 +53,31 @@ public extension InteractionContext {
...
 	}
 	
 	func adjust<AdjustableElement: Adjustable>(_ element: AdjustableElement, to newValue: AdjustableElement.Value) {
 		element.adjust(to: newValue)
 	}
 	
+	func refresh(_ refreshableThing: Refreshable) {
+		refreshableThing.refresh()
```
Comment (roger-tan) [inline TABTestKit/Classes/Contexts/InteractionContext.swift:80]:
```suggestion
        refreshableThing.refresh()
```

File: TABTestKit/Classes/Contexts/NavigationContext.swift [hunk 9]
```diff
@@ -28,7 +28,7 @@ public extension NavigationContext {
 	///
 	/// - Parameter element: The element to await.
 	func see(_ element: Element) {
-			element.await(.exists, .visible)
+		element.await(.exists, .visible)
```
Comment (roger-tan) [inline TABTestKit/Classes/Contexts/NavigationContext.swift:31]:
```suggestion
        element.await(.exists, .visible)
```

File: TABTestKit/Classes/Contexts/NavigationContext.swift [hunk 10]
```diff
@@ -58,43 +58,43 @@ public extension NavigationContext {
...
+		Icon.testRunner.tap()
+		sleep(1)
+		UIApplication.shared.open(url)
+		let openAppConfirmationAlert = Alert(id: "“\(Bundle.main.appName)” wants to open “\(App.shared.name)”", parent: springboard)
+		if openAppConfirmationAlert.determine(.exists, timeout: 1) {
+			sleep(1)
+			openAppConfirmationAlert.actionButton(withID: "Open").tap()
+		}
+	}
+
```
Comment (roger-tan) [inline TABTestKit/Classes/Contexts/NavigationContext.swift:99]:
```suggestion
    }
    
    /// Opens the specified URL from the test runner app.
    /// This means you can use this function for testing deep linking, as well as opening web urls in Safari.
    /// This function will first show the Home Screen if necessary, launch the test runner (after finding the correct
    /// page with the runner) and then open the URL.
    ///
    /// It is your responsibility/choice if you want to leave your app in the current state before the test runner
    /// opens the URL, or whether you want to completely terminate your app before hand.
    ///
    /// This leaves you free to test deep linking behaviour from the current state of the app.
    /// or a fresh launch of the app.
    ///
    /// Since you can pass any URL to this function, you could pass an app-specific url like `myapp://deep-link`, or
    /// a regular HTTP link like `https://apple.com`.
    ///
    /// HTTP links are opened in Safari by default, unless you have an app installed that can intercept them using
    /// Universal Links.
    ///
    /// - Parameter url: The URL to open.
    func open(_ url: URL) {
        springboard.activate()
        if !Icon.testRunner.determine(.hittable, timeout: 1) {
            XCUIDevice.shared.press(.home) // Ensure we're on the first page of the home screen
            sleep(1)
            while !Icon.testRunner.determine(.hittable, timeout: 1) {
                springboard.swipeLeft()
            }
        }
        Icon.testRunner.tap()
        sleep(1)
        UIApplication.shared.open(url)
        let openAppConfirmationAlert = Alert(id: "“\(Bundle.main.appName)” wants to open “\(App.shared.name)”", parent: springboard)
        if openAppConfirmationAlert.determine(.exists, timeout: 1) {
            sleep(1)
            openAppConfirmationAlert.actionButton(withID: "Open").tap()
        }
    }
    
```

File: TABTestKit/Classes/Protocols/Refreshable.swift [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+//
+
+import Foundation
+
+/// Represents something that is refreshable.
+/// Typically you'd make a screen that can be refreshed conform to this protocol,
+/// and then perform whatever needs to be performed for a "happy path" for that screen.
+public protocol Refreshable {
+	
+	func refresh()
```
Comment (roger-tan) [inline TABTestKit/Classes/Protocols/Refreshable.swift:15]:
```suggestion
    func refresh()
```

**Comment by KaneCheshire** [comment]
Please make sure the changelog is updated with any changes pending before a release :) I've been meaning to get round to creating a PR checklist with that so its harder for us to forget.
